db/hints: Log when ignoring invalid hint directories

dawmd opened the pull request:
In 58784cd, aa4b06a and other commits migrating hinted handoff from IPs to host IDs (scylladb/scylladb#15567), we started ignoring hint directories of invalid names, i.e. those that represent neither an IP address, nor a host ID. They remain on disk and are taken into account while computing, e.g. the total size of hints, but they're not used in any way.

These changes add logs informing the user when Scylla encounters such a directory.

mykaul commented:
Please add tests.

dawmd commented:
Tests verifying the changes: scylladb/scylla-dtest/pull/4006

Version 2 (I had to re-force-pushed the changes. I don't know why some other commits got involved):
* coroutinization of `space_watchdog::scan_one_ep_dir`
* taking invalid directories into account as well when computing the total space consumed by hint directories

piodul commented on db/hints/manager.cc:
```diff
@@ -185,15 +185,22 @@ future<> manager::start(shared_ptr<gms::gossiper> gossiper_ptr) {
+            try {
+                return endpoint_id{de.name};
+            } catch (...) {
+                manager_logger.warn("Encountered a hint directory of invalid name: {}. It will be ignored.", de.name);
```
> `"Encountered a hint directory of invalid name: {}. It will be ignored."`

I'd concentrate more on the consequences to the user, i.e. what it means that the directory will be ignored. In addition to the current message, I'd stress that hints from this directory will not be replayed.

piodul commented on db/hints/manager.cc:
```diff
@@ -185,15 +185,22 @@ future<> manager::start(shared_ptr<gms::gossiper> gossiper_ptr) {
     co_await lister::scan_dir(_hints_dir, lister::dir_entry_types::of<directory_entry_type::directory>(),
             [this] (fs::path datadir, directory_entry de) {
-        endpoint_id ep = endpoint_id{de.name};
+        const auto maybe_ep = std::invoke([&] () -> std::optional<endpoint_id> {
+            try {
+                return endpoint_id{de.name};
+            } catch (...) {
+                manager_logger.warn("Encountered a hint directory of invalid name: {}. It will be ignored.", de.name);
+                return {};
+            }
+        });
```
Ok, you convinced me that it's not necessarily a good idea. This is a tiny duplication that appears in a very small number of places, so we can live with it, and the alternative I proposed is not necessarily better, so let's keep it as it is.

dawmd commented:
Version 3:
* split the changes into more but single-purpose commits
* made the logs more descriptive

piodul commented on db/hints/resource_manager.cc:
```diff
@@ -142,14 +146,25 @@ void space_watchdog::on_timer() {
+                    try {
+                        return internal::endpoint_id{de.name};
+                    } catch (...) {
+                        resource_manager_logger.warn("Encountered a hint directory of invalid name: {}, exception: {}. ",
+                                de.name, std::current_exception());
```
AFAIK `space_watchdog::on_timer()` runs every second. This will generate a lot of spam. I don't think we should print it with such high verbosity, let's reduce it to `debug` - the user will know that there are issues with these directories because of the `warn` messages in `manager::start`.

dawmd replied:
That's a good point, thank you

dawmd commented:
Version 4:
* reduced the log level in `space_watchdog::on_timer()` to debug

piodul commented:
@dawmd now that https://github.com/scylladb/scylladb/pull/15567 is merged, can this PR be closed? If so then please go ahead and close it.

piodul commented:
> @dawmd now that #15567 is merged, can this PR be closed? If so then please go ahead and close it.

@dawmd ping

dawmd commented:
Version 4:
* dropped most of the commits as the changes were implemented as part of scylladb/scylladb#15567
* rebased on top of the current master
* the remaining changes of the previous version of this PR are logs. We'll use them for testing too

piodul commented:
@dawmd backport candidate label is not being honored by the "PR require backport label" check. You need to choose labels with specific versions (like e.g. `backport/6.0`) or `backport/none` if no backport is needed.

I'll add `backport/none` as this PR just adds more log messages.

piodul commented on db/hints/resource_manager.cc:
```diff
@@ -137,7 +137,8 @@ void space_watchdog::on_timer() {
         for (manager& shard_manager : per_device_limits.managers) {
             shard_manager.clear_eps_with_pending_hints();
-            lister::scan_dir(shard_manager.hints_dir(), lister::dir_entry_types::of<directory_entry_type::directory>(), [this, &shard_manager] (fs::path dir, directory_entry de) {
+            lister::scan_dir(shard_manager.hints_dir(), lister::dir_entry_types::of<directory_entry_type::directory>(),
+                    [this, &shard_manager] (fs::path dir, directory_entry de) {
```
This is a change that is completely unrelated to adding logs. I suggest dropping it.

piodul reviewed (commented):
Just one nitpick, otherwise it LGTM

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9486/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9486/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9927/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1902/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9486/)
#### Build Details:
- Duration: 6 hr 2 min
- Builder: i-0d24526d92c133953 (m5ad.8xlarge)

dawmd commented:
Version 5:
* Dropped the unnecessary indentation change

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9536/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9536/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/9970/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/1944/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/9536/)
#### Build Details:
- Duration: 2 hr 37 min
- Builder: spider5.cloudius-systems.com

piodul commented:
Queued
